### PR TITLE
KAFKA-15441 Allow broker heartbeats to complete in metadata transaction

### PR DESCRIPTION
**Pull request by mumrah**
This patch allows broker heartbeat events to be completed while a metadata transaction is in-flight.

While a metadata transaction is in progress, events added to the purgatory will not be completed because the "stable" offset will not be advancing. For ZK migrations, the transaction can be very large and take up to several minutes to complete (for very large clusters). This was causing heartbeats to sit on the purgatory and not get completed before the session expiration. This resulted in the controller fencing all the ZK brokers which has undesirable downstream effects.

### CODE REVIEW

**Review by akhileshchg — APPROVED** [review]
Looks good to me. Just to ensure this is the only issue for broker fencing, did you check that the new test fails without the fix?

**Comment by mumrah** [comment]
@akhileshchg thanks for taking a look. Yes the new test times out with the old code (since the future never completes)

**Comment by cmccabe** [comment]
Thinking about this more, If something runs in premigration it should complete in premigration. But maybe not complete inside a transaction more generally. Like I can imagine us doing something later where a heartbeat kicks off a big transaction containing a lot of records. We wouldn't want to complete inside that transaction.

tl;dr We can probably just get rid of RUNS_IN_TRANSACTION and have it be implied by:
* you have RUNS_IN_PREMIGRATION and
* we're in premigration

**Comment by mumrah** [comment]
I think that's reasonable, @cmccabe. I was looking through usages of `RUNS_IN_PREMIGRATION` and noticed we have it on `rescheduleMaybeFenceStaleBrokers`. I think we should probably remove this usage since we don't want the controller fencing any brokers while the migration is in-flight. 

I'll remove `COMPLETES_IN_TRANSACTION` and update the PR shortly.

**Review by cmccabe — APPROVED** [review]
LGTM. Thanks @mumrah

**Comment by mumrah** [comment]
~Only three flaky tests~

Edit: spoke too soon.. Still a good run

image
